api/classes/class-maybe-show-campaign.php

```diff
@@ -0,0 +1,167 @@
+<?php
+/**
+ * Newspack Campaigns maybe display campaign.
+ *
+ * @package Newspack
+ */
+
+/**
+ * Extend the base Lightweight_API class.
+ */
+require_once 'class-lightweight-api.php';
+
+/**
+ * GET endpoint to determine if campaign is shown or not.
+ */
+class Maybe_Show_Campaign extends Lightweight_API {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+		$should_show = $this->should_campaign_be_shown();
+		if ( $should_show || $this->should_perform_utm_logic() ) {
+			if ( $this->should_suppress_because_utm_suppression() ) {
+				$should_show = false;
+			}
+			if ( $this->should_suppress_because_utm_medium() ) {
+				$should_show = false;
+			}
+			if ( $should_show && $this->should_suppress_because_newsletter_campaign_dismissed() ) {
+				$show_show = false;
+			}
+		}
+
+		$this->response['displayPopup'] = $should_show;
+
+		// TODO: Preview handling.
+
+		$this->respond();
+	}
+
+	/**
+	 * Primary campaign visibility logic.
+	 *
+	 * @return bool Whether campaign should be shown.
+	 */
+	public function should_campaign_be_shown() {
+		$data                = $this->get_transient( $this->transient_name() );
+		$current_views       = (int) $data['count'];
+		$suppress_forever    = ! empty( $data['suppress_forever'] ) ? (int) $data['suppress_forever'] : false;
+		$mailing_list_status = ! empty( $data['mailing_list_status'] ) ? (int) $data['mailing_list_status'] : false;
+		$last_view           = ! empty( $data['time'] ) ? (int) $data['time'] : 0;
+
+		if ( $suppress_forever ) {
+			return false;
+		}
+		if ( $mailing_list_status ) {
+			return false;
+		}
+		if ( 'always' === $this->frequency || 'test' === $this->frequency ) {
+			return true;
+		}
+		if ( 'daily' === $this->frequency ) {
+			return $last_view < strtotime( '-1 day' );
+		}
+		if ( 'once' === $this->frequency ) {
+			return $current_views < 1;
+		}
+		return false;
+	}
+
+	/**
+	 * Assess whether the UTM logic is necessary.
+	 *
+	 * @return bool Whether UTM logic should be performed.
+	 */
+	public function should_perform_utm_logic() {
+		if ( $this->utm_suppression && stripos( urldecode( $this->referer_url ), 'utm_source=' . $this->utm_suppression ) ) {
+			return true;
+		}
+		if ( stripos( $this->referer_url, 'utm_medium=email' ) ) {
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Handle utm_source query param, update transients if needed.
+	 *
+	 * @return bool Should campaign be shown.
+	 */
+	public function should_suppress_because_utm_suppression() {
+		if ( ! $this->utm_suppression ) {
+			return false;
+		}
+		// Suppressing based on UTM Source parameter in the URL.
+		// If the visitor came from a campaign with suppressed utm_source, then it should not be displayed.
+		$utm_source_transient_name = $this->get_suppression_data_transient_name( 'utm_source' );
+		$utm_source_transient      = $this->get_transient( $utm_source_transient_name );
```

Comment:
Maybe I missed how this is handled, but don't all the `get_transient` calls in this file need checks like this?
```
$utm_source_transient      = $this->get_transient( $utm_source_transient_name );
if ( ! $utm_source_transient || ! is_array( $utm_source_transient ) ) {
    $utm_source_transient = [];
}
```

Reply:
This should be taken care of as of https://github.com/Automattic/newspack-popups/pull/208/commits/aac241f2d83fc5688b32f5d286fa5b4931fb91a7#diff-bf20f4bd9a3d050c28b9648d1c89c0d3R242. I don't think we need the `is_array` check because the DB result could only be an array or `null`.